File: typescript/src/did.ts
import { hexToBytes } from './crypto.js';

export interface DIDDocument {
  id: string;
  verificationMethod: VerificationMethod[];
  authentication: string[];
  raw?: Record<string, unknown>;
}

export interface VerificationMethod {
  id: string;
  type: string;
  controller: string;
  publicKeyMultibase?: string;
  publicKeyBytes?: string;
  publicKeyBase58?: string;
}

// Base58 decode (minimal implementation for did:key)
const BASE58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

function base58Decode(str: string): Uint8Array {
  const bytes: number[] = [0];
  for (const char of str) {
    const idx = BASE58_ALPHABET.indexOf(char);
    if (idx < 0) throw new Error(`Invalid base58 character: ${char}`);
    let carry = idx;
    for (let j = 0; j < bytes.length; j++) {
      carry += bytes[j] * 58;
      bytes[j] = carry & 0xff;
      carry >>= 8;
    }
    while (carry > 0) {
      bytes.push(carry & 0xff);
      carry >>= 8;
    }
  }
  // Leading zeros
  for (const char of str) {
    if (char !== '1') break;
    bytes.push(0);
  }
  return new Uint8Array(bytes.reverse());
}

/**
 * Resolve a did:key (Ed25519, z6Mk prefix).
 */
export function resolveDIDKey(did: string): DIDDocument {
  if (!did.startsWith('did:key:z6Mk')) {
    throw new Error(`unsupported did:key format: ${did}`);
  }

  const multibaseValue = did.split(':')[2];
  // Remove 'z' prefix (base58btc)
  const rawBytes = base58Decode(multibaseValue.slice(1));

  if (rawBytes.length < 34 || rawBytes[0] !== 0xed || rawBytes[1] !== 0x01) {
    throw new Error('invalid Ed25519 multicodec prefix');
  }

  const pubKeyBytes = rawBytes.slice(2);
  const bytesHex = Array.from(pubKeyBytes).map(b => b.toString(16).padStart(2, '0')).join('');

  const vm: VerificationMethod = {
    id: did + '#key-1',
    type: 'Ed25519VerificationKey2020',
    controller: did,
    publicKeyMultibase: multibaseValue,
    publicKeyBytes: bytesHex,
  };

  return {
    id: did,
    verificationMethod: [vm],
    authentication: [did + '#key-1'],
  };
}

/**
 * Resolve a did:web by fetching did.json.
 */
export async function resolveDIDWeb(did: string): Promise<DIDDocument> {
  if (!did.startsWith('did:web:')) {
    throw new Error(`not a did:web: ${did}`);
  }

  const parts = did.split(':').slice(2);
  const domain = parts[0].replace(/%3A/g, ':');
  const pathParts = parts.slice(1);

  const url = pathParts.length > 0
    ? `https://${domain}/${pathParts.join('/')}/did.json`
    : `https://${domain}/.well-known/did.json`;

  const resp = await fetch(url, { headers: { Accept: 'application/json' } });
  if (!resp.ok) throw new Error(`failed to fetch ${url}: ${resp.status}`);
  const raw = await resp.json() as Record<string, unknown>;

  return {
    id: (raw.id as string) || did,
    verificationMethod: (raw.verificationMethod as VerificationMethod[]) || [],
    authentication: (raw.authentication as string[]) || [],
    raw,
  };
}

/**
 * Auto-dispatch DID resolution by method.
 */
export async function resolve(did: string): Promise<DIDDocument> {
  if (did.startsWith('did:key:')) return resolveDIDKey(did);
  if (did.startsWith('did:web:')) return resolveDIDWeb(did);
  throw new Error(`unsupported DID method: ${did}`);
}

/**
 * Extract the first Ed25519 public key bytes from a DID document.
 */
export function extractPublicKey(doc: DIDDocument): Uint8Array {
  for (const vm of doc.verificationMethod) {
    if (vm.publicKeyBytes) {
      return hexToBytes(vm.publicKeyBytes);
    }
    if (vm.publicKeyMultibase) {
      const raw = base58Decode(vm.publicKeyMultibase.slice(1));
      if (raw.length >= 34 && raw[0] === 0xed && raw[1] === 0x01) {
        return raw.slice(2);
      }
      return raw;
    }
    if (vm.publicKeyBase58) {
      return base58Decode(vm.publicKeyBase58);
    }
  }
  throw new Error('no Ed25519 public key found in DID document');
}

File: python/aps/did.py
"""DID resolution for the Agent Passport Standard."""
from __future__ import annotations

import json
import urllib.request
from dataclasses import dataclass, field
from typing import Any


@dataclass
class DIDDocument:
    id: str = ""
    verification_method: list[dict[str, Any]] = field(default_factory=list)
    authentication: list[str] = field(default_factory=list)
    raw: dict[str, Any] = field(default_factory=dict)


def resolve_did_key(did: str) -> DIDDocument:
    """Resolve a did:key (Ed25519 only, z6Mk prefix)."""
    if not did.startswith("did:key:z6Mk"):
        raise ValueError(f"unsupported did:key format: {did}")

    # Extract multibase-encoded key (z = base58btc)
    import base58
    multibase_value = did.split(":", 2)[2]
    # Remove 'z' prefix (base58btc indicator)
    raw_bytes = base58.b58decode(multibase_value[1:])
    # First two bytes are multicodec prefix for Ed25519 pub key (0xed, 0x01)
    if len(raw_bytes) < 34 or raw_bytes[0] != 0xED or raw_bytes[1] != 0x01:
        raise ValueError("invalid Ed25519 multicodec prefix")
    pub_key_bytes = raw_bytes[2:]

    vm = {
        "id": did + "#key-1",
        "type": "Ed25519VerificationKey2020",
        "controller": did,
        "publicKeyMultibase": multibase_value,
        "publicKeyBytes": pub_key_bytes.hex(),
    }

    return DIDDocument(
        id=did,
        verification_method=[vm],
        authentication=[did + "#key-1"],
    )


def resolve_did_web(did: str) -> DIDDocument:
    """Resolve a did:web by fetching the did.json document."""
    if not did.startswith("did:web:"):
        raise ValueError(f"not a did:web: {did}")

    # did:web:example.com -> https://example.com/.well-known/did.json
    # did:web:example.com:path:to -> https://example.com/path/to/did.json
    parts = did.split(":")[2:]
    domain = parts[0].replace("%3A", ":")
    path_parts = parts[1:] if len(parts) > 1 else []

    if path_parts:
        url = f"https://{domain}/{'/'.join(path_parts)}/did.json"
    else:
        url = f"https://{domain}/.well-known/did.json"

    req = urllib.request.Request(url, headers={"Accept": "application/json"})
    with urllib.request.urlopen(req, timeout=10) as resp:
        raw = json.loads(resp.read())

    return DIDDocument(
        id=raw.get("id", did),
        verification_method=raw.get("verificationMethod", []),
        authentication=raw.get("authentication", []),
        raw=raw,
    )


def resolve(did: str) -> DIDDocument:
    """Auto-dispatch DID resolution by method."""
    if did.startswith("did:key:"):
        return resolve_did_key(did)
    elif did.startswith("did:web:"):
        return resolve_did_web(did)
    else:
        raise ValueError(f"unsupported DID method: {did}")


def extract_public_key(doc: DIDDocument) -> bytes:
    """Extract the first Ed25519 public key bytes from a DID document."""
    for vm in doc.verification_method:
        # From did:key resolution
        if "publicKeyBytes" in vm:
            return bytes.fromhex(vm["publicKeyBytes"])
        # From did:web / standard format
        if "publicKeyMultibase" in vm:
            import base58
            raw = base58.b58decode(vm["publicKeyMultibase"][1:])
            if len(raw) >= 34 and raw[0] == 0xED and raw[1] == 0x01:
                return raw[2:]
            return raw
        if "publicKeyBase58" in vm:
            import base58
            return base58.b58decode(vm["publicKeyBase58"])
    raise ValueError("no Ed25519 public key found in DID document")
